File: CSharp/5kyu/MemoizedFibonacci.cs
using System.Linq;
using System;
using System.Numerics;
using System.Collections.Generic;

namespace CSharp._5kyu
{
    public static partial class Kata5
    {
        public static BigInteger MemoizedFibonacci(BigInteger n)
        {
            var cashe = new Dictionary<BigInteger, BigInteger>();

            BigInteger Traverse(BigInteger n)
            {
                if (cashe.ContainsKey(n)) return cashe[n];
                if (n == 0 || n == 1) return n;

                var result = Traverse(n - 1) + Traverse(n - 2);
                cashe[n] = result;

                return result;
            }

            return Traverse(n);
        }
    }
}

File: JavaScript/5kyu/MemoizedFibonacci.js
const fibonacci = (num) => {
  var cashe = {};

  function traverse(num) {
    if (cashe[num]) return cashe[num];

    if (num == 0 || num == 1) return num;

    let result = traverse(num - 1) + traverse(num - 2);

    cashe[num] = result;

    return result;
  }

  return traverse(num);
};

console.log(fibonacci(70));
